Editorial review: Document the SpeechRecognition quality option (#44245)
mdn/content: files/en-us/web/api/speechrecognition/available_static/index.md

{{APIRef("Web Speech API")}}{{SeeCompatTable}}

> The **`available()`** static method of the [Web Speech API](/en-US/docs/Web/API/Web_Speech_API) checks whether specified languages and quality level are available for speech recognition.

Reviewer: I think "the" makes it easier to read.

```suggestion
The **`available()`** static method of the [Web Speech API](/en-US/docs/Web/API/Web_Speech_API) checks whether the specified languages and quality level are available for speech recognition.
```

Author: That's weird. I already have "the" on the interface landing page, but not on the method page. Fixed now.

  - : An object specifying options for the availability check. Possible properties include:
    - `langs`
      - : An array of one or more strings containing {{glossary("BCP 47 language tag", "BCP 47 language tags")}}, each representing a language to check for availability. Passing an empty `langs` array will not throw an error, but the return value will always resolve to `unavailable`.
    - `quality` {{optional_inline}}
      - : An enumerated value that indicates the approximate use case of your speech recognition app, and therefore, the required level of complexity of the language pack and speech recognition service. Possible values are:
        - `command`
          - : Level 1: Short isolated phrases with limited vocabulary and a single speaker. Use cases include voice commands for apps. This is the default value.
        - `dictation`
          - : Level 2: Continuous speech with moderate background noise and a single primary speaker. Use cases include dictating long-form text inputs such as SMS messages, email bodies, or strings for translation.
        - `conversation`

>           - : Level 3: Continuous speech with complex vocabulary, high background noise tolerance, and multiple primary speakers. Use cases include transcribing meetings and continuous live captioning.

Reviewer: I'd be tempted to remove the Levels here if you're not going to order alphabetically. If you are, then perhaps keep them. The levels are pretty obvious.

Author: I have thought about this comment for quite a while and decided to keep the levels, at least for now. Three reasons:

- They are mentioned in the spec, and possibly elsewhere too, so I thought they might have some kind of significance that meant it would be good to keep them.
- If a reader were to skim the list really quickly, they would see two descriptions starting with "Continuous speech..." and might think, "Oh, what's the difference"? I feel like the levels help to cement that there is a difference and make them more likely to read more carefully.
- The levels also make it quite clear why they are not in alphabetical order, which might help to guard against future hasty edits.

Reviewer: Thanks, I'm not sure I fully agree, but really I just wanted you to consider this, which you have.

Possible values include:

- `available`

>   - : Indicates that support for all the specified languages is available, at the specified `quality` level.

Reviewer: This feels a little unwieldy, perhaps because fo the double specified.

```suggestion
  - : Indicates that all the specified languages are supported at the given `quality` level.
```

Author: Good call; updated

    - If `processLocally` is set to `true`, `available` means that speech recognition is available for those languages on-device (the required language packs have been downloaded and installed on the user's computer).
    - If `processLocally` is set to `false`, `available` means that speech recognition is available for those languages either on-device or remotely.
- `downloading`

>   - : Indicates that support for the specified languages is available on-device, at the specified `quality` level, and the relevant language pack for at least one language is in the process of being downloaded. Only relevant when `processLocally` is `true`.

Reviewer: The first clause feels ambiguous (even if it is not) - it feels like we're saying its on device but then we're still downloading.

```suggestion
  - : Indicates that all the specified languages are supported at the given `quality` level for on-device processing, and that the relevant language pack for at least one language is in the process of being downloaded.
    Only relevant when `processLocally` is `true`.
```

Author: updated

- `downloadable`
  - : Indicates that support for the specified languages is available on-device, at the specified `quality` level, but the relevant language pack for at least one language has not yet been downloaded. Only relevant when `processLocally` is `true`.
- `unavailable`

>   - : Indicates that support for at least one of the specified languages is not available, at the specified `quality` level.

Reviewer: ```suggestion
  - : Indicates that at least one of the specified languages is not supported at the given `quality` level.
```

Also consider "Indicates that at least one of the specified languages is not supported, or not supported at the given `quality` level."

Author: Updated

### Checking on-device model capabilities

The following code snippet is a modification of the previous example in which we call the `available()` method with the `quality` option set to `dictation`, to check whether on-device recognition will support this quality level. If the result returned is `unavailable`, we set the `SpeechRecognition` object's {{domxref("SpeechRecognition.processLocally", "processLocally")}} property to `false` (assuming it was previously set to `true`) to force the API to use a cloud recognition service, then `start()` the recognition service.

```js
startBtn.addEventListener("click", () => {
  // Check availability of on-device target language dictation quality
  SpeechRecognition.available({
    langs: ["en-US"],
    processLocally: true,
    quality: "dictation",
  }).then((result) => {
    if (result === "unavailable") {
      diagnostic.textContent = `On-device recognition for dictation not available, running with cloud recognition`;
      recognition.processLocally = false;
      recognition.start();

>     }

Reviewer 2: nit: maybe add a comment or something for what should happen if result is anything other than `unavailable`?

Author: Good call. I've fixed this by duplicating the `quality` example from the `install()` page on the `available()` page.
